# Machine Learning March Major Project

In [194]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
np.set_printoptions(threshold=np.inf)

In [6]:
from sklearn.preprocessing import LabelEncoder

In [7]:
df=pd.read_csv("D:\\sk majorproject\\adult.csv")
df.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [8]:
df.shape

(32560, 15)

In [9]:
df.columns

Index(['39', ' State-gov', ' 77516', ' Bachelors', ' 13', ' Never-married',
       ' Adm-clerical', ' Not-in-family', ' White', ' Male', ' 2174', ' 0',
       ' 40', ' United-States', ' <=50K'],
      dtype='object')

In [10]:
duplicate = df[df.duplicated()]

In [11]:
duplicate.shape

(24, 15)

In [12]:
df.drop_duplicates(keep=False,inplace=True)

In [13]:
df.shape

(32513, 15)

</br>
</br>
</br>

## 1. Renaming the column names

In [14]:
df.columns= ['Age','Workclass','Fnlwgt','Education','Education_num','Marital_status','Occupation','relationship','race','sex', 'capital_gain','capital_loss', 'hours_per_week' ,'native_country', 'income']
df.head()

Age          Workclass  Fnlwgt   Education  Education_num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   
3   28            Private  338409   Bachelors             13   
4   37            Private  284582     Masters             14   

        Marital_status          Occupation    relationship    race      sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40            Cuba   <=50K  
4             0             0              40   United-States   <=50K

#### Adding the previous column names as first row

In [15]:
new_row = pd.DataFrame({'Age':39,'Workclass':'State-gov','Fnlwgt':'77516','Education':'Bachelors','Education_num':13,'Marital_status':'Never-married','Occupation':'Adm-clerical','relationship':'Not-in-family','race':'White','sex':'Male', 'capital_gain':2174,'capital_loss':0, 'hours_per_week':40 ,'native_country':'United-States', 'income':'<=50K'}, index =[0])
 
df = pd.concat([new_row, df[:]]).reset_index(drop = True)
df.head(5)

Age          Workclass  Fnlwgt   Education  Education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        Marital_status          Occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

</br>
</br>
</br>


## 2. Handling Null values</br>



In [16]:
df.isnull().sum()

Age               0
Workclass         0
Fnlwgt            0
Education         0
Education_num     0
Marital_status    0
Occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

</br>


In [17]:
df.columns

Index(['Age', 'Workclass', 'Fnlwgt', 'Education', 'Education_num',
       'Marital_status', 'Occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'income'],
      dtype='object')

## Column - Age

In [18]:
df['Age'].unique()

array([39, 50, 38, 53, 28, 37, 49, 52, 31, 42, 30, 23, 32, 40, 34, 25, 43,
       54, 35, 59, 56, 19, 20, 45, 22, 48, 21, 24, 57, 44, 41, 29, 18, 47,
       46, 36, 79, 27, 67, 33, 76, 17, 55, 61, 70, 64, 71, 68, 66, 51, 58,
       26, 60, 90, 75, 65, 77, 62, 63, 80, 72, 74, 69, 73, 81, 78, 88, 82,
       83, 84, 85, 86, 87], dtype=int64)

#### Changing Age column into AgeRange</br>

In [19]:
bins = [0, 2, 18, 35, 65, np.inf]
names = ['<2', '2-18', '18-35', '35-65', '65+']

df['AgeRange'] = pd.cut(df['Age'],bins,labels=names)

In [20]:
df.head()

Age          Workclass  Fnlwgt   Education  Education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        Marital_status          Occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income AgeRange  
0          2174             0              40   United-States   <=50K    35-65  
1             0             0              13   United-States   <=50K    35-65  
2             0             0              40   United-States   <=50K    35-65  
3             0             0              40   United-States   <=50K    35-65  
4             0             0              40            Cuba   <=50K    18-35

In [21]:
df.drop(['Age'], axis = 1,inplace=True)

In [22]:
lb=LabelEncoder()
df['AgeRange']=lb.fit_transform(df['AgeRange'])
df['AgeRange'].value_counts()

2    16462
0    13951
3     1156
1      945
Name: AgeRange, dtype: int64

### Column - Workclass

In [23]:
df['Workclass'].unique()

array(['State-gov', ' Self-emp-not-inc', ' Private', ' State-gov',
       ' Federal-gov', ' Local-gov', ' ?', ' Self-emp-inc',
       ' Without-pay', ' Never-worked'], dtype=object)

In [24]:
df['Workclass'] = df['Workclass'].replace([' State-gov'],'State-gov')

In [25]:
df['Workclass'] = df['Workclass'].replace([' ?'],'private')

In [26]:
df['Workclass'].unique()

array(['State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', 'private', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)

In [27]:

df['Workclass']=lb.fit_transform(df['Workclass'])
df['Workclass'].value_counts()

3    22651
5     2539
1     2093
8     1836
7     1298
4     1116
0      960
6       14
2        7
Name: Workclass, dtype: int64

### Column - Fnlwgt

In [28]:
df['Fnlwgt'].unique()


array(['77516', 83311, 215646, ..., 34066, 84661, 257302], dtype=object)

In [29]:
len(df['Fnlwgt'].unique())

21638

In [30]:
df['Fnlwgt']=pd.to_numeric(df['Fnlwgt'])

In [31]:
df['Fnlwgt'].dtype

dtype('int64')

In [32]:

df['Fnlwgt']=lb.fit_transform(df['Fnlwgt'])
df['Fnlwgt'].value_counts()

8875     13
5814     13
13070    13
4984     12
5673     12
         ..
15210     1
19000     1
6999      1
18795     1
16521     1
Name: Fnlwgt, Length: 21638, dtype: int64

### Column - Education

In [33]:
df['Education'].unique()

array(['Bachelors', ' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' Assoc-voc', ' 7th-8th',
       ' Doctorate', ' Prof-school', ' 5th-6th', ' 10th', ' 1st-4th',
       ' Preschool', ' 12th'], dtype=object)

In [34]:
df['Education'] = df['Education'].replace(['Bachelors'],' Bachelors')

In [35]:
df['Education'].unique()

array([' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' Assoc-voc', ' 7th-8th',
       ' Doctorate', ' Prof-school', ' 5th-6th', ' 10th', ' 1st-4th',
       ' Preschool', ' 12th'], dtype=object)

In [36]:
df['Education'].unique()

array([' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' Assoc-voc', ' 7th-8th',
       ' Doctorate', ' Prof-school', ' 5th-6th', ' 10th', ' 1st-4th',
       ' Preschool', ' 12th'], dtype=object)

In [37]:
df['Education'].value_counts()

 HS-grad         10487
 Some-college     7273
 Bachelors        5351
 Masters          1721
 Assoc-voc        1382
 11th             1175
 Assoc-acdm       1067
 10th              933
 7th-8th           644
 Prof-school       576
 9th               514
 12th              433
 Doctorate         413
 5th-6th           331
 1st-4th           165
 Preschool          49
Name: Education, dtype: int64

In [38]:
df['Education']=lb.fit_transform(df['Education'])
df['Education'].value_counts()

11    10487
15     7273
9      5351
12     1721
8      1382
1      1175
7      1067
0       933
5       644
14      576
6       514
2       433
10      413
4       331
3       165
13       49
Name: Education, dtype: int64

### Column - Education_num

In [39]:
df['Education_num'].unique()

array([13,  9,  7, 14,  5, 10, 12, 11,  4, 16, 15,  3,  6,  2,  1,  8],
      dtype=int64)

In [40]:

df['Education_num']=lb.fit_transform(df['Education_num'])
df['Education_num'].value_counts()

8     10487
9      7273
12     5351
13     1721
10     1382
6      1175
11     1067
5       933
3       644
14      576
4       514
7       433
15      413
2       331
1       165
0        49
Name: Education_num, dtype: int64

### Column - Marital_status

In [41]:
df['Marital_status'].unique()

array(['Never-married', ' Married-civ-spouse', ' Divorced',
       ' Married-spouse-absent', ' Never-married', ' Separated',
       ' Married-AF-spouse', ' Widowed'], dtype=object)

In [42]:
df['Marital_status'] = df['Marital_status'].replace(['Never-married'],' Never-married')

In [43]:
df['Marital_status'].unique()

array([' Never-married', ' Married-civ-spouse', ' Divorced',
       ' Married-spouse-absent', ' Separated', ' Married-AF-spouse',
       ' Widowed'], dtype=object)

In [44]:
df['Marital_status']=lb.fit_transform(df['Marital_status'])
df['Marital_status'].value_counts()

2    14964
4    10652
0     4439
5     1025
6      993
3      418
1       23
Name: Marital_status, dtype: int64

### Column - relationship

In [45]:
df['relationship'].unique()

array(['Not-in-family', ' Husband', ' Not-in-family', ' Wife',
       ' Own-child', ' Unmarried', ' Other-relative'], dtype=object)

In [46]:
df['relationship'] = df['relationship'].replace(['Not-in-family'],' Not-in-family')

In [47]:
df['relationship'].unique()

array([' Not-in-family', ' Husband', ' Wife', ' Own-child', ' Unmarried',
       ' Other-relative'], dtype=object)

In [48]:
df['relationship']=lb.fit_transform(df['relationship'])
df['relationship'].value_counts()

0    13181
1     8280
3     5060
4     3444
5     1568
2      981
Name: relationship, dtype: int64

### Column - race

In [49]:
df['race'].unique()

array(['White', ' White', ' Black', ' Asian-Pac-Islander',
       ' Amer-Indian-Eskimo', ' Other'], dtype=object)

In [50]:
df['race'] = df['race'].replace(['White'],' White')

In [51]:
df['race'].unique()

array([' White', ' Black', ' Asian-Pac-Islander', ' Amer-Indian-Eskimo',
       ' Other'], dtype=object)

In [52]:
df['race']=lb.fit_transform(df['race'])
df['race'].value_counts()

4    27775
2     3120
1     1037
0      311
3      271
Name: race, dtype: int64

### Column - sex

In [53]:
df['sex'].unique()

array(['Male', ' Male', ' Female'], dtype=object)

In [54]:
df['sex'] = df['sex'].replace(['Male'],' Male')

In [55]:
df['sex'].unique()

array([' Male', ' Female'], dtype=object)

In [56]:
df['sex']=lb.fit_transform(df['sex'])
df['sex'].value_counts()

1    21760
0    10754
Name: sex, dtype: int64

### Column - capital_gain

In [57]:
df['capital_gain'].value_counts()

0        29802
15024      347
7688       284
7298       246
99999      159
         ...  
1111         1
2538         1
22040        1
4931         1
5060         1
Name: capital_gain, Length: 119, dtype: int64

In [58]:
df['capital_gain']=lb.fit_transform(df['capital_gain'])
df['capital_gain'].value_counts()

0      29802
108      347
94       284
91       246
118      159
       ...  
8          1
37         1
112        1
75         1
78         1
Name: capital_gain, Length: 119, dtype: int64

### Column - capital_loss

In [59]:
df['capital_loss'].value_counts()

0       30995
1902      202
1977      168
1887      159
1848       51
1485       51
2415       49
1602       47
1740       42
1590       40
1876       39
1672       34
1564       25
2258       25
2001       24
1669       24
1741       24
1980       23
1719       22
2051       21
2002       21
1408       21
1579       20
2377       20
1504       18
1721       18
1974       18
2339       17
2179       15
1628       15
1762       14
2559       12
625        12
2444       12
2824       10
1617        9
2042        9
2205        9
2392        9
1651        9
1594        8
2174        7
1380        7
1340        7
1092        7
2206        6
2057        6
880         6
1573        6
2246        6
2603        5
213         4
1825        4
1668        4
1258        4
2547        4
1726        4
2129        3
419         3
2231        3
2267        3
2457        3
323         3
653         3
4356        3
3770        2
1755        2
3683        2
1648        2
2754        2
3900        2
1138  

In [60]:
df['capital_loss']=lb.fit_transform(df['capital_loss'])
df['capital_loss'].value_counts()

0     30995
47      202
50      168
46      159
44       51
17       51
76       49
25       47
37       42
23       40
45       39
32       34
20       25
69       25
52       24
31       24
38       24
51       23
33       22
55       21
53       21
15       21
22       20
74       20
18       18
34       18
49       18
72       17
62       15
27       15
40       14
83       12
5        12
77       12
86       10
26        9
54        9
64        9
75        9
29        9
24        8
61        7
14        7
13        7
10        7
65        6
56        6
8         6
21        6
68        6
84        5
2         4
42        4
30        4
12        4
82        4
35        4
58        3
4         3
66        3
70        3
78        3
3         3
6         3
91        3
89        2
39        2
88        2
28        2
85        2
90        2
11        2
67        2
59        2
9         2
7         2
41        2
87        2
36        2
73        2
63        1
80        1
60        1
79  

### Column - hours_per_week

In [61]:
df['capital_loss'].value_counts()

0     30995
47      202
50      168
46      159
44       51
17       51
76       49
25       47
37       42
23       40
45       39
32       34
20       25
69       25
52       24
31       24
38       24
51       23
33       22
55       21
53       21
15       21
22       20
74       20
18       18
34       18
49       18
72       17
62       15
27       15
40       14
83       12
5        12
77       12
86       10
26        9
54        9
64        9
75        9
29        9
24        8
61        7
14        7
13        7
10        7
65        6
56        6
8         6
21        6
68        6
84        5
2         4
42        4
30        4
12        4
82        4
35        4
58        3
4         3
66        3
70        3
78        3
3         3
6         3
91        3
89        2
39        2
88        2
28        2
85        2
90        2
11        2
67        2
59        2
9         2
7         2
41        2
87        2
36        2
73        2
63        1
80        1
60        1
79  

In [62]:
df['hours_per_week']=lb.fit_transform(df['hours_per_week'])
df['hours_per_week'].value_counts()

39    15192
49     2815
44     1822
59     1475
34     1295
19     1224
29     1147
54      692
24      674
47      515
37      476
14      404
68      291
9       272
31      266
23      252
64      244
35      220
41      219
43      212
15      205
11      173
42      151
36      149
7       145
51      138
76      133
55       97
27       86
93       85
45       82
17       75
69       71
72       66
5        64
4        60
3        54
46       49
79       45
21       44
53       41
32       39
2        39
38       38
40       36
1        32
13       32
26       30
25       30
16       29
48       29
85       29
33       28
57       28
6        26
52       25
20       24
12       23
22       21
0        20
61       18
8        18
56       17
65       17
18       14
63       14
50       13
80       13
67       12
92       11
10       11
62       10
75        8
28        7
74        6
30        5
58        5
90        5
66        4
73        3
77        3
86        3
60        2
81  

### Column - native_country

In [63]:
df['native_country'].unique()

array(['United-States', ' United-States', ' Cuba', ' Jamaica', ' India',
       ' ?', ' Mexico', ' South', ' Puerto-Rico', ' Honduras', ' England',
       ' Canada', ' Germany', ' Iran', ' Philippines', ' Italy',
       ' Poland', ' Columbia', ' Cambodia', ' Thailand', ' Ecuador',
       ' Laos', ' Taiwan', ' Haiti', ' Portugal', ' Dominican-Republic',
       ' El-Salvador', ' France', ' Guatemala', ' China', ' Japan',
       ' Yugoslavia', ' Peru', ' Outlying-US(Guam-USVI-etc)', ' Scotland',
       ' Trinadad&Tobago', ' Greece', ' Nicaragua', ' Vietnam', ' Hong',
       ' Ireland', ' Hungary', ' Holand-Netherlands'], dtype=object)

In [64]:
df['native_country'].value_counts()

 United-States                 29135
 Mexico                          635
 ?                               581
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Japan                            62
 Guatemala                        61
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 

In [65]:
df['native_country'] = df['native_country'].replace([' ?'],' United-States')

In [66]:
df['native_country'].value_counts()

 United-States                 29716
 Mexico                          635
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Japan                            62
 Guatemala                        61
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 Greece                           29
 

In [67]:
df['native_country']=lb.fit_transform(df['native_country'])
df['native_country'].value_counts()

38    29716
25      635
29      198
10      137
1       121
32      114
7       106
18      100
4        95
8        90
22       81
34       80
2        75
21       73
5        70
39       67
23       62
12       61
30       60
3        59
35       51
13       44
19       43
31       37
26       34
28       31
11       29
9        29
6        28
20       24
16       20
37       19
0        19
36       18
24       18
40       16
27       14
17       13
15       13
33       12
41        1
14        1
Name: native_country, dtype: int64

### Column - income

In [68]:
df['income'].value_counts()

 <=50K    24676
 >50K      7837
<=50K         1
Name: income, dtype: int64

In [69]:
df['income'] = df['income'].replace(['<=50K'],' <=50K')

In [70]:
df['income'].value_counts()

 <=50K    24677
 >50K      7837
Name: income, dtype: int64

In [71]:
df['income']=lb.fit_transform(df['income'])
df['income'].value_counts()

0    24677
1     7837
Name: income, dtype: int64

### n

### Column - Occupation

In [72]:
df['Occupation'].unique()

array(['Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Adm-clerical', ' Sales',
       ' Craft-repair', ' Transport-moving', ' Farming-fishing',
       ' Machine-op-inspct', ' Tech-support', ' ?', ' Protective-serv',
       ' Armed-Forces', ' Priv-house-serv'], dtype=object)

In [73]:
df['Occupation'] = df['Occupation'].replace(['Adm-clerical'],' Adm-clerical')

In [162]:
df['Occupation'].value_counts()

 Prof-specialty       4132
 Craft-repair         4089
 Exec-managerial      4064
 Adm-clerical         3766
 Sales                3650
 Other-service        3287
 Machine-op-inspct    1998
 ?                    1843
 Transport-moving     1597
 Handlers-cleaners    1368
 Farming-fishing      990 
 Tech-support         926 
 Protective-serv      649 
 Priv-house-serv      146 
 Armed-Forces         9   
Name: Occupation, dtype: int64

### Finding the missing value(?) using RandomForestClassifier

In [75]:
df1=df.loc[df['Occupation'] == ' ?']
df2=df.loc[df['Occupation'] != ' ?']

In [76]:
df1.shape

(1843, 15)

In [77]:
df2.shape

(30671, 15)

In [78]:
df1['Occupation']=lb.fit_transform(df1['Occupation'])
df1['Occupation'].value_counts()

C:\Users\heman\AppData\Local\Temp/ipykernel_13860/1366951021.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Occupation']=lb.fit_transform(df1['Occupation'])


0    1843
Name: Occupation, dtype: int64

In [79]:
df2['Occupation']=lb.fit_transform(df2['Occupation'])
df2['Occupation'].value_counts()

C:\Users\heman\AppData\Local\Temp/ipykernel_13860/213460150.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Occupation']=lb.fit_transform(df2['Occupation'])


9     4132
2     4089
3     4064
0     3766
11    3650
7     3287
6     1998
13    1597
5     1368
4      990
12     926
10     649
8      146
1        9
Name: Occupation, dtype: int64

In [157]:
df1.columns

Index(['Workclass', 'Fnlwgt', 'Education', 'Education_num', 'Marital_status', 'Occupation', 'relationship', 'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'income', 'AgeRange'], dtype='object')

In [81]:
df1x= df1.drop('Occupation',axis=1)

In [82]:
df1x.shape

(1843, 14)

In [83]:
df1y= df1.iloc[:,5]

In [153]:
df1y

array([ 3,  2,  9, ...,  9,  7, 10])

In [85]:
df2x= df2.drop('Occupation',axis=1)

In [86]:
df2x.shape

(30671, 14)

In [87]:
df2y= df2.iloc[:,5]

In [154]:
df2y

0         0
1         3
2         5
3         5
4         9
         ..
32509    12
32510     6
32511     0
32512     0
32513     3
Name: Occupation, Length: 30671, dtype: int32

In [89]:
df2x.columns

Index(['Workclass', 'Fnlwgt', 'Education', 'Education_num', 'Marital_status',
       'relationship', 'race', 'sex', 'capital_gain', 'capital_loss',
       'hours_per_week', 'native_country', 'income', 'AgeRange'],
      dtype='object')

In [90]:
df2y

0         0
1         3
2         5
3         5
4         9
         ..
32509    12
32510     6
32511     0
32512     0
32513     3
Name: Occupation, Length: 30671, dtype: int32

In [91]:
from sklearn.ensemble import RandomForestClassifier

In [92]:
m= RandomForestClassifier()
m.fit(df2x,df2y)

RandomForestClassifier()

In [93]:
print('training Score',m.score(df2x,df2y))
df1y=m.predict(df1x)
print(df1y)

training Score 0.9957940725767012
[ 3  2  9 ...  9  7 10]


In [94]:
np.unique(df1y, return_counts=True)

(array([ 0,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13]),
 array([431, 135, 139, 106,  14,  19, 433,   2, 313,  87,  52,  30,  82],
       dtype=int64))

In [95]:
df1x.shape

(1843, 14)

In [96]:
df1y.shape

(1843,)

In [97]:
df2x.shape

(30671, 14)

In [98]:
df2y.shape

(30671,)

In [99]:
df1x['Occupation']=df1y

In [100]:
df1x.head()

Workclass  Fnlwgt  Education  Education_num  Marital_status  \
27           8   10520         15              9               2   
61           8   18032          5              3               3   
69           8   12801         15              9               4   
77           8   13861          0              5               2   
106          8   18385          0              5               4   

     relationship  race  sex  capital_gain  capital_loss  hours_per_week  \
27              0     1    1             0             0              59   
61              1     4    1             0             0              39   
69              3     4    1             0             0              39   
77              0     4    1             0             0               1   
106             3     4    0           116             0              31   

     native_country  income  AgeRange  Occupation  
27               34       1         2           3  
61               38       0         0           2  
69               38       0         0           9  
77               38       0         3           2  
106              38       0         1           7

In [101]:
df2x['Occupation']=df2y

In [102]:
df2x.head()

Workclass  Fnlwgt  Education  Education_num  Marital_status  relationship  \
0          7    2671          9             12               4             1   
1          5    2926          9             12               2             0   
2          3   14082         11              8               0             1   
3          3   15332          1              6               2             0   
4          3   19346          9             12               2             5   

   race  sex  capital_gain  capital_loss  hours_per_week  native_country  \
0     4    1            25             0              39              41   
1     4    1             0             0              12              38   
2     4    1             0             0              39              38   
3     2    1             0             0              39              38   
4     2    0             0             0              39               4   

   income  AgeRange  Occupation  
0       0         2           0  
1       0         2           3  
2       0         2           5  
3       0         2           5  
4       0         0           9

In [103]:
df1x.head()

Workclass  Fnlwgt  Education  Education_num  Marital_status  \
27           8   10520         15              9               2   
61           8   18032          5              3               3   
69           8   12801         15              9               4   
77           8   13861          0              5               2   
106          8   18385          0              5               4   

     relationship  race  sex  capital_gain  capital_loss  hours_per_week  \
27              0     1    1             0             0              59   
61              1     4    1             0             0              39   
69              3     4    1             0             0              39   
77              0     4    1             0             0               1   
106             3     4    0           116             0              31   

     native_country  income  AgeRange  Occupation  
27               34       1         2           3  
61               38       0         0           2  
69               38       0         0           9  
77               38       0         3           2  
106              38       0         1           7

In [104]:
df_final=df2x.append(df1x)

In [105]:
df_final

Workclass  Fnlwgt  Education  Education_num  Marital_status  \
0              7    2671          9             12               4   
1              5    2926          9             12               2   
2              3   14082         11              8               0   
3              3   15332          1              6               2   
4              3   19346          9             12               2   
...          ...     ...        ...            ...             ...   
32483          8   18856          9             12               2   
32484          8     673          9             12               4   
32492          8   17801         10             15               2   
32494          8   13007         11              8               5   
32495          8    6333         11              8               2   

       relationship  race  sex  capital_gain  capital_loss  hours_per_week  \
0                 1     4    1            25             0              39   
1                 0     4    1             0             0              12   
2                 1     4    1             0             0              39   
3                 0     2    1             0             0              39   
4                 5     2    0             0             0              39   
...             ...   ...  ...           ...           ...             ...   
32483             5     4    0             0             0              54   
32484             1     1    0             0             0              93   
32492             0     4    1             0             0               9   
32494             1     2    0             0             0              31   
32495             0     4    1             0             0              24   

       native_country  income  AgeRange  Occupation  
0                  41       0         2           0  
1                  38       0         2           3  
2                  38       0         2           5  
3                  38       0         2           5  
4                   4       0         0           9  
...               ...     ...       ...         ...  
32483              38       1         0           3  
32484              38       0         0           9  
32492              38       1         3           9  
32494              38       0         2           7  
32495              38       0         3          10  

[32514 rows x 15 columns]

</br>
</br>

## 3. Spliting the Dataset into training and testing data

In [106]:
from sklearn.model_selection import train_test_split

In [107]:
x=df_final.drop(['income'],axis=1)
y=df_final.iloc[:,-3]
print(x.shape)
print(y.shape)

(32514, 14)
(32514,)


In [108]:
print(type(x))
print(type(y))

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>


In [109]:
x.head()

Workclass  Fnlwgt  Education  Education_num  Marital_status  relationship  \
0          7    2671          9             12               4             1   
1          5    2926          9             12               2             0   
2          3   14082         11              8               0             1   
3          3   15332          1              6               2             0   
4          3   19346          9             12               2             5   

   race  sex  capital_gain  capital_loss  hours_per_week  native_country  \
0     4    1            25             0              39              41   
1     4    1             0             0              12              38   
2     4    1             0             0              39              38   
3     2    1             0             0              39              38   
4     2    0             0             0              39               4   

   AgeRange  Occupation  
0         2           0  
1         2           3  
2         2           5  
3         2           5  
4         0           9

In [110]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: income, dtype: int32

In [111]:
x_tr,x_te,y_tr,y_te=train_test_split(x,y,test_size=0.20)
print(x_tr.shape)
print(x_te.shape)
print(y_tr.shape)
print(y_te.shape)

(26011, 14)
(6503, 14)
(26011,)
(6503,)


</br>
</br>

## 4. Applying the classifier models on training dataset and generating predictions for the            test dataset

#### 4a. DECISIONTREE CLASSIFIER

In [112]:
from sklearn.tree import DecisionTreeClassifier

In [113]:
m1= DecisionTreeClassifier()
m1.fit(x_tr,y_tr)

DecisionTreeClassifier()

In [158]:
print('training Score',m1.score(x_tr,y_tr))
print('testing Score',m1.score(x_te,y_te))
ypred_m1=m1.predict(x_te)
print(ypred_m1)

training Score 0.999884664180539
testing Score 0.8150084576349377
[0 0 1 ... 0 0 0]


#### 4b. RANDOMFOREST CLASSIFIER

In [115]:
from sklearn.ensemble import RandomForestClassifier

In [116]:
m2= RandomForestClassifier()
m2.fit(x_tr,y_tr)

RandomForestClassifier()

In [212]:
print('training Score',m2.score(x_tr,y_tr))
print('testing Score',m2.score(x_te,y_te))
ypred_m2=m2.predict(x_te)
print(ypred_m2)

training Score 0.9997308830879243
testing Score 0.8536060279870828
[0 0 0 0 0 0 1 1 0 0 0 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1
 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 1 1 1 0 0 0 0 0 0 0
 0 1 1 0 0 0 1 0 0 1 0 0 1 0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1
 0 0 1 0 1 1 0 1 1 1 0 0 0 1 0 0 0 0 1 1 1 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0
 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 1 0 1 1 1 0 0 1 0 0 0 0 1 0 1 0 0
 1 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0
 1 0 0 0 1 0 0 0 0 0 1 0 1 0 1 0 1 1 0 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 1 1
 0 0 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0
 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 1 1 0 1 1 0 0 1 0 0 1 0 1 

#### 4c. KNN CLASSIFIER

In [118]:
from sklearn.neighbors import KNeighborsClassifier

In [119]:
m3=KNeighborsClassifier(n_neighbors=11)
m3.fit(x_tr,y_tr)

KNeighborsClassifier(n_neighbors=11)

In [192]:
print('training Score',m3.score(x_tr,y_tr))
print('testing Score',m3.score(x_te,y_te))
ypred_m3=m3.predict(x_te)
print(ypred_m3)

training Score 0.8106570297181962
testing Score 0.8016300169152699
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0
 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0
 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 1 0 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0
 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 1 

#### 4d. LOGISTIC REGRESSION

In [121]:
from sklearn.linear_model import LogisticRegression

In [191]:
m4=LogisticRegression()
m4.fit(x_tr,y_tr)

C:\Users\heman\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [190]:
print('training Score',m4.score(x_tr,y_tr))
print('testing Score',m4.score(x_te,y_te))
ypred_m4=m4.predict(x_te)
print(ypred_m4)

training Score 0.8067356118565222
testing Score 0.818545286790712
[0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 1
 1 0 0 1 0 0 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 1 0 0 0 0
 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 1 1 0 1 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1 1 0 0 0
 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0
 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0
 1 0 0 0 0 1 0 0 0 0 1 1 0 0 0 1 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0
 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0

#### 4e. SVM CLASSIFIER

In [124]:
from sklearn.svm import SVC

In [125]:
m5=SVC(kernel='linear',C=10)
m5.fit(x_tr,y_tr)

SVC(C=10, kernel='linear')

In [213]:
print('training Score',m5.score(x_tr,y_tr))
print('testing Score',m5.score(x_te,y_te))
ypred_m5=m5.predict(x_te)
print(ypred_m5)

training Score 0.8164622659643996
testing Score 0.8220821159464863
[0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0
 1 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 1 0 0 1 0 0 1 0 0 0 0 0 1 1 0 0 0
 1 0 1 0 0 0 0 0 1 1 0 0 0 1 1 1 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0
 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 1 1 0 1 0 0 1 1 0 0 0 1 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0
 0 1 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 

</br>
</br>

## 5.Predicting the income for test data

#### 5a. Decision Tree Classifier 

In [189]:
ypred_m1=m1.predict(x_te)
ypred_m1

array([0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,

#### 5b. Random Forest Classifier  

In [188]:
ypred_m2=m2.predict(x_te)
ypred_m2

array([0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,

#### 5c. KNN Classifier 

In [187]:
ypred_m3=m3.predict(x_te)
ypred_m3

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

#### 5d. Logistic Regression 

In [186]:
ypred_m4=m4.predict(x_te)
ypred_m4

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

#### 5e. SVM Classifier 

In [193]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 100)
pd.set_option('display.width', 1000)
ypred_m5=m5.predict(x_te)
print(ypred_m5)

[0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0
 1 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 1 0 0 1 0 0 1 0 0 0 0 0 1 1 0 0 0
 1 0 1 0 0 0 0 0 1 1 0 0 0 1 1 1 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0
 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 1 1 0 1 0 0 1 1 0 0 0 1 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0
 0 1 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0
 0 1 0 0 0 0 0 0 0 0 0 0 

</br>
</br>

## 6. CONFUSION_METRIX AND CLASSIFICATION_REPORT

In [132]:
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

#### 6a. Decision Tree Classifier 

In [133]:
cm_m1=confusion_matrix(y_te,ypred_m1)
print(cm_m1)
print(classification_report(y_te,ypred_m1))

[[4342  623]
 [ 580  958]]
              precision    recall  f1-score   support

           0       0.88      0.87      0.88      4965
           1       0.61      0.62      0.61      1538

    accuracy                           0.82      6503
   macro avg       0.74      0.75      0.75      6503
weighted avg       0.82      0.82      0.82      6503



#### 6b. Random Forest Classifier 

In [134]:
cm_m2=confusion_matrix(y_te,ypred_m2)
print(cm_m2)
print(classification_report(y_te,ypred_m2))

[[4582  383]
 [ 569  969]]
              precision    recall  f1-score   support

           0       0.89      0.92      0.91      4965
           1       0.72      0.63      0.67      1538

    accuracy                           0.85      6503
   macro avg       0.80      0.78      0.79      6503
weighted avg       0.85      0.85      0.85      6503



#### 6c. KNN Classifier 

In [135]:
cm_m3=confusion_matrix(y_te,ypred_m3)
print(cm_m3)
print(classification_report(y_te,ypred_m3))

[[4808  157]
 [1133  405]]
              precision    recall  f1-score   support

           0       0.81      0.97      0.88      4965
           1       0.72      0.26      0.39      1538

    accuracy                           0.80      6503
   macro avg       0.76      0.62      0.63      6503
weighted avg       0.79      0.80      0.76      6503



#### 6d. Logistic Regression 


In [136]:
cm_m4=confusion_matrix(y_te,ypred_m4)
print(cm_m4)
print(classification_report(y_te,ypred_m4))

[[4663  302]
 [ 878  660]]
              precision    recall  f1-score   support

           0       0.84      0.94      0.89      4965
           1       0.69      0.43      0.53      1538

    accuracy                           0.82      6503
   macro avg       0.76      0.68      0.71      6503
weighted avg       0.80      0.82      0.80      6503



#### 6e. SVM Classifier 


In [137]:
cm_m5=confusion_matrix(y_te,ypred_m5)
print(cm_m5)
print(classification_report(y_te,ypred_m5))

[[4646  319]
 [ 838  700]]
              precision    recall  f1-score   support

           0       0.85      0.94      0.89      4965
           1       0.69      0.46      0.55      1538

    accuracy                           0.82      6503
   macro avg       0.77      0.70      0.72      6503
weighted avg       0.81      0.82      0.81      6503



</br>
</br>

## 7. Validate the result for Precision, Recall, F1-score and Accuracy for each model based on values from confusion_matrix and classification_report

In [206]:
from sklearn.metrics import precision_score,recall_score,f1_score,accuracy_score

#### a. Decision Tree Classifier 

In [207]:
print('Precision: %.3f' % precision_score(y_te,ypred_m1))
print('Recall: %.3f' % recall_score(y_te,ypred_m1))
print('f1: %.3f' % f1_score(y_te,ypred_m1))
print('Accuracy: %.3f' % accuracy_score(y_te,ypred_m1))

Precision: 0.606
Recall: 0.623
f1: 0.614
Accuracy: 0.815


#### 7b. Random Forest Classifier 


In [208]:
from sklearn.metrics import precision_score,recall_score,f1_score,accuracy_score
print('Precision: %.3f' % precision_score(y_te,ypred_m2))
print('Recall: %.3f' % recall_score(y_te,ypred_m2))
print('f1: %.3f' % f1_score(y_te,ypred_m2))
print('Accuracy: %.3f' % accuracy_score(y_te,ypred_m2))

Precision: 0.717
Recall: 0.630
f1: 0.671
Accuracy: 0.854


#### 7c. KNN Classifier 

In [209]:
from sklearn.metrics import precision_score,recall_score,f1_score,accuracy_score
print('Precision: %.3f' % precision_score(y_te,ypred_m3))
print('Recall: %.3f' % recall_score(y_te,ypred_m3))
print('f1: %.3f' % f1_score(y_te,ypred_m3))
print('Accuracy: %.3f' % accuracy_score(y_te,ypred_m3))

Precision: 0.721
Recall: 0.263
f1: 0.386
Accuracy: 0.802


#### 7d. Logistic Regression 


In [210]:
from sklearn.metrics import precision_score,recall_score,f1_score,accuracy_score
print('Precision: %.3f' % precision_score(y_te,ypred_m4))
print('Recall: %.3f' % recall_score(y_te,ypred_m4))
print('f1: %.3f' % f1_score(y_te,ypred_m4))
print('Accuracy: %.3f' % accuracy_score(y_te,ypred_m4))

Precision: 0.686
Recall: 0.429
f1: 0.528
Accuracy: 0.819


#### 7e. SVM Classifier 


In [211]:
from sklearn.metrics import precision_score,recall_score,f1_score,accuracy_score
print('Precision: %.3f' % precision_score(y_te,ypred_m5))
print('Recall: %.3f' % recall_score(y_te,ypred_m5))
print('f1: %.3f' % f1_score(y_te,ypred_m5))
print('Accuracy: %.3f' % accuracy_score(y_te,ypred_m5))

Precision: 0.687
Recall: 0.455
f1: 0.548
Accuracy: 0.822


</br>
</br>

## 8. Generate the percentage of misclassification in each of these models</br>
$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$ Percentage of missclasification is <b>1-Accuracy<b>.

#### a. Decision Tree Classifier 


Percentage of missclasification = 1-0.815 = 0.185

#### 8b. Random Forest Classifier


Percentage of missclasification = 1-0.853 = 0.147

#### 8c. KNN Classifier

Percentage of missclasification = 1-0.801 = 0.199

#### 8d. Logistic Regression

Percentage of missclasification = 1-0.818 = 0.182

#### 8e. SVM Classifier

Percentage of missclasification = 1-0.822 = 0.178

## 9. Accuracies of all the models

a) Decision Tree Classifier  - 0.815 <br>
b) Random Forest Classifier  - 0.853 <br>
c) KNN Classifier            - 0.801 <br>
d) Logistic Regression       - 0.818 <br>
e) SVM Classifier            - 0.822

#### Therefore the RandomForest Classifier Model is having the best accuracy

</br>
</br>
</br>
</br>

In [1]:
pwd

'C:\\Users\\heman'